### ISSUE TITLE
Problem with loading DotLoader

### ISSUE THREAD

**Issue by bratislavdamnjanovic** [bhargavms/DotLoader#4]
Hi there,

We have a issue with loading the dot loader. Sometimes it crashes. Can you advise us?

Cheers,
Bratislav

# Platform: android
# OS Version: 5.0.1
# Device: GT-I9500
# RAM Free: 24%
# Disk Free: 22.8%

#0. Crashed: main: 0 0 0x0000000000000000
       at com.bhargavms.dotloader.DotLoader._startAnimation(DotLoader.java:219)
       at com.bhargavms.dotloader.DotLoader.access$200(DotLoader.java:21)
       at com.bhargavms.dotloader.DotLoader$2.run(DotLoader.java:196)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:5951)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1400)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1195)

--

Fatal Exception: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.animation.ValueAnimator.start()' on a null object reference
       at com.bhargavms.dotloader.DotLoader._startAnimation(DotLoader.java:219)
       at com.bhargavms.dotloader.DotLoader.access$200(DotLoader.java:21)
       at com.bhargavms.dotloader.DotLoader$2.run(DotLoader.java:196)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:5951)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1400)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1195)

**Comment by bhargavms** [comment]
Can you add the relevant code in here too? From what I can see in the stacktrace apparently the ValueAnimator is null, (i.e not initialised) Need to look at your code to find out why this can happen

**Comment by bhargavms** [comment]
@bratislavdamnjanovic The animations are initialized in `onMeasure()` because that is where the height of the view is calculated and the height is needed to calculate how high the dot must bounce. So is `onMeasure()` called after the `Runnable` passed to `view.post()` is executed?

**Comment by bratislavdamnjanovic** [comment]
I wil test it today.

Here is the code:

z__portable_loader.xml

<?xml version="1.0" encoding="utf-8"?>
<FrameLayout style="@style/wrap"
			 xmlns:android="http://schemas.android.com/apk/res/android"
			 android:layout_centerInParent="true">

	<com.bhargavms.dotloader.DotLoader
		android:id="@+id/loader_bar"
		style="@style/loader" />

	<TextView
		android:id="@+id/loader_txt"
		android:visibility="gone"
		style="@style/wrap"
		android:gravity="center"
		android:textColor="@color/grey_dark_primary"
		android:textSize="@dimen/txt_16" />

</FrameLayout>


public void init() {
		addView(inflate(getContext(), R.layout.z__portable_loader, null));
		mLoader = findViewById(R.id.loader_bar);
		mTextView = (TextView) findViewById(R.id.loader_txt);
	}

public void show() {
		mLoader.setVisibility(View.VISIBLE);
		mTextView.setVisibility(View.GONE);
	}

**Comment by bhargavms** [comment]
@bratislavdamnjanovic I will use your code and try to replicate tonight, you will have a fix/workaround by tomorrow if I can reproduce the issue

**Comment by bratislavdamnjanovic** [comment]
On LG Nexus 5x does not crash. For now, just the device in the first comment (Samsung GT-I9500 5.0.1).
Thank you!

**Comment by bhargavms** [comment]
@bratislavdamnjanovic Ok then its device specific, apparently Samsung GT-I9500 calls `onMeasure()` after the runnable passed to `view.post(runnable)` , which is unexpected, I will push a fix for it now, can you try it and tell me if it fixed your problem?

**Comment by bratislavdamnjanovic** [comment]
Sure. Thanks

**Comment by bhargavms** [comment]
@bratislavdamnjanovic I have released v1.0.2, please update to this version and check if the issue still exists

**Comment by bratislavdamnjanovic** [comment]
It's still the same. Same error occurred

**Comment by bratislavdamnjanovic** [comment]
Can I provide you some more information in order to analyze?

**Comment by bhargavms** [comment]
can you paste the code for `@style/loader` ?

**Comment by bhargavms** [comment]
@bratislavdamnjanovic I managed to get my hands on a Samsung GT I9500, and I have used the same code as you have done, i.e inflating and adding the view, then I have also added setVisibility GONe and VISIBLE. still unable to reproduce.

**Comment by bratislavdamnjanovic** [comment]
It happens sometimes. We caught some error on Sony e6853 also.

styles.xml 

```
<style name="loader" parent="wrap">
		<item name="android:layout_centerInParent">true</item>
		<item name="color_array">@array/dot_colors</item>
		<item name="dot_radius">7dp</item>
		<item name="number_of_dots">3</item>
</style>
```



values.xml 

```
<resources>
	<array name="dot_colors">
		<item>#e1e7eb</item>
		<item>#5B85AB</item>
		<item>#4d4d5a</item>
		<item>#ca6608</item>
	</array>
</resources>
```



layout.xml

```
<DrawerLayout>
<RelativeLayout ...>
<LinearLayout...>
<RelativeLayout
        style="@style/fill">

	<android.support.v7.widget.RecyclerView
	android:id="@+id/list"
	style="@style/match"
	android:background="@drawable/bg_white"/>

	<q4u.f4m.tc.quiz4you.utils.Loader
		android:id="@+id/loader"
		style="@style/loader" />
</RelativeLayout>
</LinearLayout>
</RelativeLayout>
<DrawerLayout>
```

**Comment by bratislavdamnjanovic** [comment]
Hi. We have a Websocket call, so basically waiting for the server response. Maybe it's happening that the response from the server is to fast, and something happen when I mLoader.setVisibility(View.GONE);? Or something earlier. I can't figure it out

**Comment by bhargavms** [comment]
So do you have the device with you? or did this issue pop on your Fabric crashlytics?

**Comment by bratislavdamnjanovic** [comment]
Samsung is here with me.

**Comment by bhargavms** [comment]
Can you put a break point in `onMeasure()` of `DotLoader` and a break point inside `_startAnimation` and tell me which one is getting called first?

**Comment by bratislavdamnjanovic** [comment]
First is called onMeasure(). But the problem is that this not happens all the time. Maybe 1 in 10 times.

**Comment by bratislavdamnjanovic** [comment]
```
 FATAL EXCEPTION: main
                                                                     Process: q4u.f4m.tc.quiz4you, PID: 12220
                                                                     java.lang.NullPointerException: Attempt to invoke virtual method 'void android.animation.ValueAnimator.start()' on a null object reference
                                                                         at com.bhargavms.dotloader.DotLoader._startAnimation(DotLoader.java:219)
                                                                         at com.bhargavms.dotloader.DotLoader.access$200(DotLoader.java:21)
                                                                         at com.bhargavms.dotloader.DotLoader$2.run(DotLoader.java:196)
                                                                         at android.os.Handler.handleCallback(Handler.java:739)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                         at android.os.Looper.loop(Looper.java:145)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5951)

```

**Comment by bhargavms** [comment]
ok, I can suggest a workaround to you, as soon as you do `mLoader.setVisibility(View.VISIBLE)` call `mLoader.forceLayout()` to trigger `onMeasure() again`

**Comment by bratislavdamnjanovic** [comment]
OK. I'll test it. Thanks!

**Comment by bratislavdamnjanovic** [comment]
Check the error from the console

**Comment by bhargavms** [comment]
@bratislavdamnjanovic yes I'm looking at it, the only way the value animator can be null is if `onMeasure()` is not called during the initialization of the view, I don't understand how that is possible.

**Comment by gish94** [comment]
i have same issue. When i return to app from recent app list. 

FATAL EXCEPTION: main
                                                                    Process: com.softsun.hurryapp, PID: 7524
                                                                    java.lang.NullPointerException: Attempt to invoke virtual method 'void android.animation.ValueAnimator.start()' on a null object reference
                                                                        at com.bhargavms.dotloader.DotLoader._startAnimation(DotLoader.java:219)
                                                                        at com.bhargavms.dotloader.DotLoader.access$200(DotLoader.java:21)
                                                                        at com.bhargavms.dotloader.DotLoader$2.run(DotLoader.java:196)
                                                                        at android.os.Handler.handleCallback(Handler.java:751)
                                                                        at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                        at android.os.Looper.loop(Looper.java:154)
                                                                        at android.app.ActivityThread.main(ActivityThread.java:6077)
                                                                        at java.lang.reflect.Method.invoke(Native Method)
                                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)

**Comment by bhargavms** [comment]
This is resolved now with the latest release version 1.0.3
